### PR TITLE
Fix workflow that checks for Redox schema changes

### PR DESCRIPTION
**Pull request by mmabey**
Previously, the `handle_updates.yml` workflow was using actions that were outdated and no longer supported by GitHub. Those actions have been fixed and the workflow now references the updated versions. There was also an issue with maintaining the proper working directory arising from the workflow performing operations in two repos in separate directories.

Note: One of the steps in this workflow is to automatically open a PR in [`cedar-team/pyredox`](https://github.com/cedar-team/pyredox) when `redox_lib_gen` detects a schema change. I haven't had an opportunity to test if this works as expected yet, so I don't know what issues there are with that (based on my reading, I expect there to be an authentication error that might be easily solved by uncommenting a couple lines at the end of the workflow file). Since I no longer have direct access to the `pyredox` repo, I won't be notified if that ever fails.

Finally, the [`hotfix/node12_to_16_pt2`](https://github.com/cedar-team/redox_lib_gen/tree/hotfix/node12_to_16_pt2) branch can be deleted now.

### CODE REVIEW

**Comment by mmabey** [comment]
@jjcollette @pjones-cedar I apologize for pinging you two about this, but I wasn't sure who else could help me with this request.

### **Please respond by COB August 25, 2023**

I am feeling a little frustrated about the situation with the `pyredox` and `redox_lib_gen` repositories. I understand that reviewing contributions to these projects is not a very high priority and I'm not asking for that to change because I know this library is only ancillary to Cedar's purpose. But, since this work was approved to be made open-source, I believe there comes with that a responsibility to outline some path for continued updates and maintenance, whether handled internally or delegated to some outside third party.

I request that, for the benefit of the `pyredox` library, **one** of two actions takes place:
1. Cedar makes me a collaborator on the `pyredox` and `redox_lib_gen` repositories so that I can continue to make updates, corrections, and other necessary code changes.
2. Cedar transfers ownership of the `pyredox` and `redox_lib_gen` repositories to me.

My reasons for this request are as follows:
- Given the company's past behavior related to the `pyredox` library, I don't foresee Cedar dedicating resources to reviewing PRs and resolving issues in a timely manner any time in the near future.
- Up until 6 weeks ago, the person at Cedar most qualified and most likely to review PRs and issues was me. Now that I'm gone, `pyredox` and `redox_lib_gen` now have exactly zero code contributors remaining at Cedar. While I would be happy to see this change, there again seems to be a lack of alignment between the library's needs and Cedar's priorities.
- Multiple opportunities already exist for improving the library. As things currently stand, I cannot add these opportunities to any kind of roadmap for future versions. Absent the permissions of a repository collaborator or owner, I also cannot create documentation so that other members of the open source community may contribute work towards these efforts.


I believe the risk to Cedar from either of the proposed courses of action are low since the main cedar repo is pinned to a version of `pyredox` that was created while I was still an employee there. Also, since it will continue to be an open-source project, Cedar can inspect its contents before deciding to use new versions of the library.

I don't mean to sound ungrateful for the sponsorship Cedar provided for me to write this library, I just don't want to see it languish from neglect. I want to be a good steward of this cool library I made.

### **Please respond to my request by COB August 25, 2023** (at which point this untouched PR will be more than a month old).
If I don't hear anything by then, I will assume Cedar isn't interested in retaining ownership of the library and I will take the necessary steps to protect its future.

File: .github/workflows/handle_updates.yml [hunk 1]
```diff
@@ -9,48 +9,38 @@ jobs:
...
       # Check out and set up redox_lib_gen
-      - name: Checkout
+      - name: Checkout lib generator
         uses: actions/checkout@v3
         with:
           path: redox_lib_gen
 
       - name: Install Python and Poetry
-        uses: ITProKyle/action-setup-python@v0.3.0
+        uses: mmabey/action-setup-python@v0.3.1
```
Comment (pjones-cedar) [inline .github/workflows/handle_updates.yml:18]: https://github.com/actions/setup-python

Would prefer the official setup-python action

**Review by pjones-cedar — CHANGES_REQUESTED** [review]
- [ ] I fully understand what the code is doing and its context
- [ ] The code is the best way to accomplish this problem (I would code it similarly)
- [ ] The code is well organized, easy to understand and matches Cedar design patterns and best practices
- [ ] The code handles edge cases (that I can think of) and has tests to validate them
- [ ] The appropriate people are reviewing (knowledge and security)
- [ ] I tested the code personally (Not required)

File: .github/workflows/handle_updates.yml (hunk 1 above)
Comment (mmabey) [reply .github/workflows/handle_updates.yml:18]: Thanks for taking a look Pen, I really appreciate it.

The official setup-python action is fine, but since Poetry also needs to be set up with the same version of Python, I was happy when I found an action that did both in one go. Once GitHub dropped legacy support for Node v12, the `ITProKyle/action-setup-python` action stopped working completely, which is what prompted a lot of the work in this PR.

The code of [my version of the action](https://github.com/mmabey/action-setup-python/blob/master/action.yml) does exactly 5 things:
1. Uses the official setup-python action
2. Sets up Poetry
3. Uses a cache for the virtual environment
4. Checks that the cache is still healthy
5. Installs the Poetry virtual environment and the packages as specified in the lock file

If you feel strongly about only using the official setup-python action, I can find a workaround. But I think that having those steps in their own action makes sense. And the only reason I started maintaining my own version of that was because the original author wasn't merging the Dependabot PRs.

Hopefully that context is helpful. Please let me know your thoughts!

**Comment by taylor-cedar** [comment]
Hey @mmabey! Hope you are doing well. I really appreciate you bringing this to our attention. I am really sorry this has been an issue. We unfortunately don't have enough resources to dedicate to maintaining this library actively. We can only update when it's necessary. 

We can help getting this current issue fixed, but going forward we won't be accepting contributions outside of Cedar. I will update the README.md to make this clear.

If you would prefer to more actively manage it, I encourage you to fork this code and release it yourself. We use a very permissive license to allow anyone to use it however they want.

**Comment by mmabey** [comment]
@taylor-cedar Thanks for the response. To make sure I understand you correctly, please let me repeat back what your decision is. Cedar is unable maintain the library, but won't accept contributions from non-Cedar personnel. Cedar is also uninterested in transferring ownership of the library's GitHub repo to someone that will provide ongoing maintenance. Is that correct?

I'll have to assume there's more to your reasoning here, because it doesn't make sense to me with the details you've shared. Whatever your reasons may be, I hope Cedar will take them into consideration for future proposals to make libraries like this open-source. It doesn't seem fair to Cedar or the open-source community for the long-term maintenance plan to be the continued employment of a single person.

This particular PR only fixes a problem with the auto-generation of the `pyredox` library in the event that Redox publishes changes to their schema. From my perspective, if Cedar cannot devote resources to reviewing the auto-generated PR that would result from such an event, there's no longer a need to fix that feature. So, I'll close this PR with a heavy heart knowing that `pyredox`, like its creator, is "no longer worth Cedar's investment."
